intern.go

```diff
@@ -71,74 +70,109 @@ func Get(cmpVal interface{}) *Value {
 		v = valSafe[cmpVal]
 	} else if addr, ok := valMap[cmpVal]; ok {
 		v = (*Value)((unsafe.Pointer)(addr))
-	}
-	if v != nil {
-		v.refs++
+		// We've materialized a *Value out of thin air, in violation
+		// of Go's unsafety rules. The GC won't "see" this pointer
+		// until the next GC cycle. This is also true of normal
+		// pointers. Unlike normal pointers however, it's possible
+		// that the GC has just finished a mark phase in which it
+		// concluded that v was unreferenced, and thus collectable.
+		//
+		// By setting resurrected here, we signal to v's finalizer
+		// that the Value needs to remain alive for at least one more
+		// GC cycle, long enough for the GC to become aware of v's
+		// existence. See the comments in finalize for the other leg
+		// of this interaction.
+		v.resurrected = true
 		return v
 	}
-	v = &Value{cmpVal: cmpVal, refs: 1}
-	if valSafe != nil {
-		valSafe[cmpVal] = v
-	} else {
-		valMap[cmpVal] = uintptr(unsafe.Pointer(v))
-		runtime.SetFinalizer(v, finalize)
-	}
+	// When we first create a value, v is a "correct" pointer from the
+	// runtime's perspective. There is no risk of concurrent
+	// collection, so we can set resurrected=false.
+	v = &Value{cmpVal: cmpVal, resurrected: false}
+	valMap[cmpVal] = uintptr(unsafe.Pointer(v))
+	runtime.SetFinalizer(v, finalize)
 	return v
 }
 
 func finalize(v *Value) {
 	mu.Lock()
 	defer mu.Unlock()
-	v.refs--
-	if v.refs == 0 {
+
+	// Now that the finalizer is holding mu, we're sure that Get
+	// cannot unsafely materialize more references to v. But maybe it
+	// already did materialize an unsafe reference previously?
+	if v.resurrected {
+		// We may have lost a race with Get, and there may be a new
+		// pointer to v that the GC is unaware of. We can't let it
+		// delete v. To prevent that, we reinstall ourselves as
+		// finalizer, which will make the value persist until the next
+		// GC cycle. We also clear the sentinel, so that if there
+		// truly are no more references to v, the next finalizer run
+		// will clean up.
+		//
+		// Note that this check is not precise. The call to Get that
+		// set resurrected may have happened thousands of GC cycles
+		// ago, and the GC may be entirely correct that v is
+		// collectable. But we can't know for certain until we got a
+		// chance to run and clear resurrected, so even in the
+		// non-racy case, Values that have been used more than once
+		// take 2 GC cycles to actually be freed.
+		v.resurrected = false
+		runtime.SetFinalizer(v, finalize)
+	} else {
+		// There have been no new pointers to v materialized since the
```

Comment:
slight preference for keeping the early return and outdenting this path as I did in my concurrent #7 

because the `if v.resurrected` is essentially an error path. The normal flow should be at the same indent level as above.